test/unit/puppetlabs/puppetserver/lockable_pool_test.clj

```diff
@@ -60,6 +85,27 @@
         ;; make sure we got here
         (is (true? true))))))
 
+(deftest pool-lock-is-blocking-test-until-borrows-unregistered-test
+  (let [pool (create-populated-pool 3)
+        instances (borrow-n-instances pool 2)
+        future-started? (promise)]
+    (is (= 2 (count instances)))
+    (let [lock-thread (future (deliver future-started? true)
+                              (.lock pool))]
+      @future-started?
+      (testing "pool.lock() blocks until all instances are unregistered"
+        (is (not (realized? lock-thread)))
+
+        (testing (str "other threads may successfully unregister instances "
+                      "while pool.lock() is being executed")
+          (.unregister pool (first instances))
+          (is (not (realized? lock-thread)))
+          (.unregister pool (second instances)))
+
+        @lock-thread
+        ;; make sure we got here
+        (is (true? true))))))
```

Comment:
Yeah, that's what I was getting at, is wondering what would happen if the lock was requested and then the last instance was unregistered.  It seems like the only case where that could currently happen is what you're describing, where we're trying to lock the old pool while a full pool flush is in progress, and so the lock should probably fail when the last instance is unregistered... but maybe it needs to happen via some kind of exception that we can catch and recognize so that we can retry the lock acquisition?